src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1279,18 +1280,39 @@ internal void FinishRead(byte[] renegotiateBuffer)
             }
         }
 
-        // Returns: 
-        // true  - operation queued
-        // false - operation can proceed
-        internal bool CheckEnqueueWrite(AsyncProtocolRequest asyncRequest)
+        internal Task CheckEnqueueWriteAsync()
+        {
+            //Clear previous request.
+            int lockState = Interlocked.CompareExchange(ref _lockWriteState, LockWrite, LockNone);
+            if (lockState != LockHandshake)
+            {
+                return Task.CompletedTask;
+            }
+
+            lock (this)
+            {
+                if (_lockWriteState != LockHandshake)
```

Comment:
Is there a race condition here?  I realize this predates your PR, but it looks strange.  Outside of the lock the field is being manipulated by an interlocked, and then here there's a check protected by a lock, but that lock and interlocked have nothing to do with each other... could another thread be doing that interlocked operation at the same time we're checking/setting _lockWriteState here?